File: Lab 7/ProblemB.py
import heapq
import sys

def dijkstra(start_node, adj_list, n_var):
    INF = 10**18
    dist = [INF] * n_var
    dist[start_node] = 0

    pq = []
    heapq.heappush(pq, (0, start_node))

    while pq:
        cur_dist, node = heapq.heappop(pq)
        if cur_dist > dist[node]:
            continue

        for nxt, wgt in adj_list[node]:
            nd = cur_dist + wgt
            if nd < dist[nxt]:
                dist[nxt] = nd
                heapq.heappush(pq, (nd, nxt))

    return dist


def where_to_meet():
    data = sys.stdin.read().split()
    idx = 0

    n_var = int(data[idx]); idx += 1
    m_var = int(data[idx]); idx += 1
    a_var = int(data[idx]) - 1; idx += 1
    b_var = int(data[idx]) - 1; idx += 1

    adj_list = [[] for _ in range(n_var)]

    i_var = 0
    while i_var < m_var:
        u = int(data[idx]) - 1
        v = int(data[idx + 1]) - 1
        w = int(data[idx + 2])
        idx += 3
        adj_list[u].append((v, w))
        i_var += 1

    dist_a = dijkstra(a_var, adj_list, n_var)
    dist_b = dijkstra(b_var, adj_list, n_var)

    INF = 10**18
    best_time = INF
    best_node = -1

    i_var = 0
    while i_var < n_var:
        if dist_a[i_var] != INF and dist_b[i_var] != INF:
            meet_time = max(dist_a[i_var], dist_b[i_var])
            if meet_time < best_time or (meet_time == best_time and i_var < best_node):
                best_time = meet_time
                best_node = i_var
        i_var += 1

    if best_node == -1:
        print(-1)
    else:
        print(best_time, best_node + 1)

File: Lab 7/Java/ProblemB.java
import java.util.*;

public class ProblemB {
    static class Node implements Comparable<Node> {
        int v;
        long w;

        public Node(int v, long w) {
            this.v = v;
            this.w = w;
        }

        public int compareTo(Node other) {
            return Long.compare(this.w, other.w);
        }
    }

    static long[] dijkstra(int start, List<List<Node>> adj, int n) {
        long INF = (long) 1e18;
        long[] dist = new long[n];
        Arrays.fill(dist, INF);
        dist[start] = 0;
        PriorityQueue<Node> pq = new PriorityQueue<>();
        pq.add(new Node(start, 0));

        while (!pq.isEmpty()) {
            Node curr = pq.poll();
            if (curr.w > dist[curr.v])
                continue;
            for (Node neighbor : adj.get(curr.v)) {
                if (dist[curr.v] + neighbor.w < dist[neighbor.v]) {
                    dist[neighbor.v] = dist[curr.v] + neighbor.w;
                    pq.add(new Node(neighbor.v, dist[neighbor.v]));
                }
            }
        }
        return dist;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextInt())
            return;
        int n = sc.nextInt();
        int m = sc.nextInt();
        int a = sc.nextInt() - 1;
        int b = sc.nextInt() - 1;

        // ProblemB.py reads: u, v, w sequentially in loop.
        // "i_var < m_var: u = ... v = ... w = ..."
        // This is simpler structure.

        List<List<Node>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++)
            adj.add(new ArrayList<>());

        for (int i = 0; i < m; i++) {
            int u = sc.nextInt() - 1;
            int v = sc.nextInt() - 1;
            int w = sc.nextInt();
            adj.get(u).add(new Node(v, w));
        }

        long[] distA = dijkstra(a, adj, n);
        long[] distB = dijkstra(b, adj, n);

        long INF = (long) 1e18;
        long bestTime = INF;
        int bestNode = -1;

        for (int i = 0; i < n; i++) {
            if (distA[i] != INF && distB[i] != INF) {
                long meetTime = Math.max(distA[i], distB[i]);
                if (meetTime < bestTime) {
                    bestTime = meetTime;
                    bestNode = i;
                } else if (meetTime == bestTime) {
                    if (bestNode == -1 || i < bestNode)
                        bestNode = i;
                }
            }
        }

        if (bestNode == -1)
            System.out.println("-1");
        else
            System.out.println(bestTime + " " + (bestNode + 1));
    }
}
